0005-kzen-two-party-ecdsa.adoc

```diff
@@ -0,0 +1,182 @@
+Practically, this means you don't need to lock Bitcoin to multiple public keys and check the signature with `OP_CHECKMULTISIG`; you can just lock it to a joint public key using a normal `P2WSH` output (i.e. use `OP_CHECKSIG`).
+There's no way to tell the difference between a single user `P2WSH` output and a `P2WSH` output from a joint public key.
+
+Since ECDSA signatures don't have a nice algebraic structure it is **very** difficult to do key aggregation.
+In schnorr you can quite literally just add each party's share of the signature together.
+The method we have to use in ECDSA is to move the operations to produce the *s* component of the signature into a https://en.wikipedia.org/wiki/Paillier_cryptosystem[Paillier group] rather than doing them simply modulo the Secp256k1 curve order.
+Once you're in Paillier land things become much easier but it comes at a cost of boatloads of messages and complexity during the key generation and exchange phase.
+It is so complicated that it's amazing that anyone actually bothered to come up with this.
+For us it may be very useful.
+Even though Bitcoin will get Schnorr signatures, Ethereum probably won't switch anytime soon.
+
+The hard bit is doing the setup phase which involves quite a few proofs.
+Some of them are in the above crate but some of them are imported from elsewhere.
+We'll start with going through these.
+
+== Key Generation
+
+=== Generating the joint public key
+
+To produce signatures under a joint public key you first need to generate it fairly.
+Because of rogue key attacks, you can't just have one party send their public key and the other send one back.
+
+The technique used in the paper (and followed by the implementation) is to do a coin tossing protocol to determine the joint public key.
+This ensures that the public key is totally random and cannot be biased by the party that chooses their key afterwards.
+Note, The ability to simulate a random coin toss gives the key generation very strong security guarantees.
+
+The protocol is roughly as follows, where x~1~ is P1's (resp. P2) private key and X~1~ = g^x~1~^  is her public key.
+
+- P1 -> P2: Hash(X~1~, schnorr_nizk(X~1~))
+- P2 -> P1: X~2~, schnorr_nizk(X~2~)
+- P1 -> P2: X~1~, schnorr_nizk(X~1~) (i.e. opens the commitment from 1)
+- Both parties calculate the joint public key X = X~1~^x~2~^ = X~2~^x~1~^ (i.e. Diffie-Hellman)
+
+==== Implementation notes
+The implementation commits by sending two hashes, one for the key and one for the NIZK.
+
+The NIZK is implemented here: https://github.com/KZen-networks/curv/blob/18f0081a9a3025eafd49ad496e5032647edf4aa3/src/cryptographic_primitives/proofs/sigma_dlog.rs
+
+=== Proving Paillier modulus is well formed
+
+In order to use Paillier encryption P1 needs to generate a RSA type modulus `N = pq` where `p` and `q` are large primes.
+P1 needs to prove that she generated `N` properly (I don't actually know what the risk is to P2 if she doesn't).
```

Comment:
According to the paper this is needed in order to proof that `gcd(N, φ(N) = 1` holds and hence is indeed additively homomorphic.

Reply:
But what does this mean for P2? Does P2 leak secret information somewhere if it's not? Or maybe P2 will accept an invalid proof from P1 if this doesn't hold. I'm sure there's a reason but didn't come across it the paper (yet!).